### PR TITLE
V3 reorganization 

### PR DESCRIPTION
**Pull request by mcuadros**
- `billy.Filesystem` has been split in several dedicated interfaces, also the test suite has been divided in several suites.
-  subdirfs & tmpfs: converted to helpers, this fs are not anymore handled as fs, since they are not.
-  memfs & osfs: based on helper chroot, simplifying a lot the code and make it less error prone.
- `helper.Chroot` has being covered by a dedicated test suite to make it easier detected issues.

### CODE REVIEW

File: fs.go [hunk 1]
```diff
@@ -4,82 +4,151 @@ import (
...
 	"io"
 	"os"
+	"time"
 )
 
 var (
-	ErrClosed       = errors.New("file: Writing on closed file.")
-	ErrReadOnly     = errors.New("this is a read-only filesystem")
-	ErrNotSupported = errors.New("feature not supported")
+	ErrReadOnly        = errors.New("read-only filesystem")
```
Comment (smola) [inline fs.go]: Should we  use `gopkg.in/src-d/go-errors.v0` now that we're moving to v3?

File: fs.go [hunk 2]
```diff
@@ -4,82 +4,151 @@ import (
...
+	// It is the caller's responsibility to remove the file when no longer
+	// needed.
+	TempFile(dir, prefix string) (File, error)
 }
 
-type FileInfo os.FileInfo
+// Dir abstract the dir related operations in a storage-agnostic interface as
+// an extension to the Basic interface.
+type Dir interface {
+	ReadDir(path string) ([]os.FileInfo, error)
```
Comment (smola) [inline fs.go]: Add doc? Even if it's just `the same as the standard library`

File: fs.go [hunk 3]
```diff
@@ -4,82 +4,151 @@ import (
...
+	// Readlink returns the target path of link. If there is an error, it will
+	// be of type *PathError.
+	Readlink(link string) (string, error)
 }
 
-func (f *BaseFile) Filename() string {
-	return f.BaseFilename
+// Change abstract the FileInfo change related operations in a storage-agnostic
+// interface as an extension to the Basic interface
+type Change interface {
```
Comment (smola) [inline fs.go]: `Change` is too broad. Maybe `AttributeChanger` or something like that?

File: fs.go [hunk 4]
```diff
@@ -4,82 +4,151 @@ import (
...
+}
 type removerAll interface {
 	RemoveAll(string) error
 }
+
+// File represent a file, being a subset of the os.File
+type File interface {
+	// Name returns the name of the file as presented to Open.
+	Name() string
+	io.Writer
```
Comment (smola) [inline fs.go]: `io.WriterAt`?

File: helper/mount/mount.go [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+
+	uc capabilities // underlying
+	sc capabilities // source
+}
+
+type capabilities struct{ dir, symlink bool }
+
+// New creates a new filesystem wrapping up 'fs' the intercepts all the calls
+// made to `mountpoint` path and redirecting it to `source` filesystem.
+func New(fs billy.Basic, mountpoint string, source billy.Basic) *Mount {
```
Comment (smola) [inline helper/mount/mount.go]: Neat.

File: memfs/memory.go [hunk 6]
```diff
@@ -11,52 +11,46 @@ import (
...
-//New returns a new Memory filesystem
-func New() *Memory {
-	return &Memory{
-		base: string(separator),
-		s:    newStorage(),
-	}
+//New returns a new Memory filesystem.
+func New() billy.Filesystem {
+	fs := &Memory{s: newStorage()}
+	return chroot.New(fs, string(separator))
```
Comment (smola) [inline memfs/memory.go]: Why wrapping it with Chroot?
Maybe it should just implement the `Chroott` function and use `chroot.New` to implement it? Or is this to save the definition of `Chroot` completely?

File: utils.go [hunk 7]
```diff
@@ -5,20 +5,27 @@ import (
...
+// can but returns the first error it encounters. If the path does not exist,
+// RemoveAll returns nil (no error).
+func RemoveAll(fs Basic, path string) error {
+	u, ok := fs.(underlying)
 	if ok {
+		if ch, ok := fs.(Chroot); ok {
+			path = fs.Join(ch.Root(), path)
+		}
+
+		fs = u.Underlying()
```
Comment (smola) [inline utils.go]: :+1:

File: memfs/memory.go (hunk 6 above)
Comment (mcuadros) [inline memfs/memory.go]: because you need to intercept all the call to provide the real path.

File: fs.go (hunk 4 above)
Comment (mcuadros) [inline fs.go]: yep... maybe, is more constant, but then we need to implement :p

File: fs.go (hunk 3 above)
Comment (mcuadros) [inline fs.go]: yep, maybe

File: fs.go (hunk 2 above)
Comment (mcuadros) [inline fs.go]: sure

File: fs.go (hunk 1 above)
Comment (mcuadros) [inline fs.go]: yep, and no. I need to fit the documentation and this means, use os.PathError and os.LinkError in all the function to mimic the standard library

**Pull request merged**